### PR TITLE
Create ergodatetime.

### PR DESCRIPTION
**Pull request by Anjani-225**
<feat>(<extraction -> stdlib -> ergodatetime>): <new date time library writtern completely in ergo language ,writtern so as to save maximum gas when deployed on any chain in the future> - <Issue #345 -implement DateTime within Ergo>

# Issue #345 
<Implementation of a new library of DateTime within ergo. Created in accordance with the comments given in the issues.>

### Changes
 -ergodatetime.ergo
 - in ergo language
 - accounts for leap year problem
 - accounts for checking the weekend as these contracts are mainly legal 
 - accounts for gas costs as it does not include any outside libraries.

### Flags
- Unsure about the functions that return multiple values

### Related Issues
Issue #345 – Implement datetime within ergo

### CODE REVIEW

**Comment by Anjani-225** [comment]
@irmerk I have implemented the changes. I am a little confused as to why the DCO check is still unsuccessful. Could you please take a look?

**Comment by mttrbrts** [comment]
> @irmerk I have implemented the changes. I am a little confused as to why the DCO check is still unsuccessful. Could you please take a look?

Hi @Anjani-225 , the DCO log here (https://github.com/accordproject/ergo/pull/740/checks?check_run_id=525990576) shows that this commit is missing the sign-off, https://github.com/accordproject/ergo/pull/740/commits/8e862f314819fde24558e89ffa4b83666c9e0c14.

File: extraction/stdlib/ergodatetime. [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+//made on template studio
+namespace org.accordproject.ergodatetime
+
+define constant SECONDS_PER_DAY = 24 * 60 * 60
+define constant SECONDS_PER_HOUR = 60 * 60
+define constant SECONDS_PER_MINUTE = 60
+define constant OFFSET = 2440588  //so that the secods may start counting from 0(ist jan 1970 is 0)
+//assignng number to each day of the weekDay
+
+define constant MON = 1
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: Why not use an enum?
```
define enum {
  MON,
  TUE,
...}
```

File: extraction/stdlib/ergodatetime. [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//made on template studio
+namespace org.accordproject.ergodatetime
+
+define constant SECONDS_PER_DAY = 24 * 60 * 60
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: the other way around might be better?
```
define constant SECONDS_PER_MINUTE = 60
define constant SECONDS_PER_HOUR = 60 * SECONDS_PER_MINUTE
...
```

File: extraction/stdlib/ergodatetime. [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+define constant TUE = 2
+define constant WED = 3
+define constant THU = 4
+define constant FRI = 5
+define constant SAT = 6
+define constant SUN = 7
+
+ /*
+ the date conversion algorithm from
+ http://aa.usno.navy.mil/faq/docs/JD_Formula.php
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: I can't find that link?

File: extraction/stdlib/ergodatetime. [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ */
+
+  define function DayToDate(year : Double , month : Double , day : Double) : Double {
+        enforce year >= 1970;
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: Why can't we do `year < 1970` ?

File: extraction/stdlib/ergodatetime. [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    
+       let days : Double = day
+          - 32075
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: Where do those constant come from? What do they mean?

File: extraction/stdlib/ergodatetime. [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+          + 1461 * (year + 4800 + (month - 14) / 12) / 4
+          + 367 * (month - 2 - (month - 14) / 12 * 12) / 12
+          - 3 * ((year + 4900 + (month - 14) / 12) / 100) / 4
+          - OFFSET;
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: Where is `OFFSET` defined?

File: extraction/stdlib/ergodatetime. [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+
+    
+    function leapyearcheck(timestamp : Double)  leapYear : Boolean {
+        (uint year,,) = daysToDate(timestamp / SECONDS_PER_DAY);
+        leapYear = _isLeapYear(year);
+    }
+   
+
+	function getDayOfWeek(uint timestamp) internal pure returns (uint dayOfWeek) {
```
Comment (jeromesimeon) [inline extraction/stdlib/ergodatetime.]: Wrong syntax? did you mean `timestamp:Double ... : Double`? What's `internal pure` ? Was this code taken from somewhere else, if yes, make sure to document it in the issue or PR.

Comment (Anjani-225) [reply extraction/stdlib/ergodatetime.]: Yes the original code was in solidity as I am not as familiar with Ergo as of now

**Comment by Anjani-225** [comment]
Thank you. I will rectify the errors ASAP!

File: extraction/stdlib/ergodatetime. (hunk 7 above)
Comment (jeromesimeon) [reply extraction/stdlib/ergodatetime.]: > Yes the original code was in solidity as I am not as familiar with Ergo as of now

oh ok great! Is that code open-source? A pointer would help please.

**Comment by jeromesimeon** [comment]
@Anjani-225 Any further interest in this?

**Comment by jeromesimeon** [comment]
This is stale, so closing. Feel free to reopen if any new things!

**Pull request closed without merging**